src/main/kotlin/com/trackforever/controllers/ProjectController.kt

```diff
@@ -96,58 +103,132 @@ class ProjectController (@Autowired private var projectRepository: ProjectReposi
+            successes.isEmpty() -> ResponseEntity(HttpStatus.GONE) // None of the projects were found.
+            successes.keys.size == content.keys.size -> ResponseEntity(successes, HttpStatus.OK) // All projects were found
+            else -> ResponseEntity(successes, HttpStatus.MULTI_STATUS)
         }
     }
 
     /**
      * Adds the given projects to the database.
      * Expects a List of projects in a JSON String.
-     * Possible Responses: HTTP 200 OK upon completion.
-     * TODO: Verify that the projects being passed in are valid
+     * Possible Responses: HTTP 200 OK upon completion with a List of projects
+     *                     that were successfully saved.
      */
     @CrossOrigin
     @PutMapping("/projects")
-    fun setProjects(@RequestBody projects: String): ResponseEntity<String> {
+    fun setProjects(@RequestBody projects: String): ResponseEntity<List<TrackForeverProject>> {
         logger.debug("set projects :: ($projects)")
         val mapper = jacksonObjectMapper()
         val content: List<TrackForeverProject> = mapper.readValue(projects)
+        val successes: MutableList<TrackForeverProject> = mutableListOf()
         // Add to the database
         content.forEach {
-            projectRepository.save(it)
+            // Look in the database for the previous project.
+            val serverProject: TrackForeverProject
+            try { // Project with a prevHash found
+                serverProject = projectRepository.findByHash(it.prevHash)
+                // Delete the previous one and insert the new one after checking its current hash
+                val calculatedHash = generateHash(it)
+                if (calculatedHash != it.hash) {
+                    logger.debug("Mismatched project hash detected.")
+                    logger.debug("Given hash: ${it.hash}")
+                    logger.debug("Expected hash: $calculatedHash")
+                    it.hash = calculatedHash
+                }
+                projectRepository.deleteById(serverProject.id)
+                successes += it
+                it.issues.forEach {
+                    val calcIssueHash = generateHash(it.value)
+                    if (calcIssueHash != it.value.hash) { // Fix hashes for issues
+                        logger.debug("Mismatched issue hash detected.")
+                        logger.debug("Given hash: ${it.value.hash}")
+                        logger.debug("Expected hash: $calcIssueHash")
+                        it.value.hash = calcIssueHash
+                    }
+                }
+                projectRepository.save(it)
+            } catch (e: EmptyResultDataAccessException) { // No project with such a prevHash found.
+                // Check the newly calculated hash and add it to the repository
+                val calculatedHash = generateHash(it)
+                if (calculatedHash != it.hash) { // Fix hashes for projects
```

Comment:
This won't work. The hash will be wrong if the user has made changes. The local hash doesn't update until it's been synced with the server.